### PR TITLE
Add sampling memory profiler mode

### PR DESCRIPTION
**Pull request by travisdowns**
This series changes the existing heap profiling to support a sampling mode, added
by @StephanDollberg and includes also some optimizations of the allocation fast paths.

The following description of the sampling profiler comes from Stephan:

Instead of tracking every allocation, it will track a sampled subset of allocations. This
is useful for profiling in production systems where the overhead of tracking every
allocation is too high.

A sample is taken at a set sample rate which feeds into an exponential
distribution. Hence we avoid falling into patterns that would potentially make
the profile biased otherwise. This approach is inspired by the android heapprofd
memory profiler.

By adjusting the sample rate it's possible to tune the overhead of the memory
profiler. The idea is to make this an always-on production profiler that can
give insight into allocations in OOM scenarios with negliable overhead.

Starting with the latest release of Redpanda we are using this in a default
always-on mode with a sampling period of 3MB. This is a rather conservative
setting and results in no measurable overhead while still giving good resolution
for the top allocation contributors.

We also add an API to retrieve the current heap profile programmatically. In
Redpanda we use this to print the top sites on OOM using the additional
diagnostics API. Further it allows easier retrieval of the data in a
dev/profiling fashion.

As today heap profiler can be compiled in or not (using `--heap-prof` at 
./configure.py time) and can also be enabled or disabled at runtime if it
has been compiled in.

The individual commits contain a more detailed description of the respective
changes. Note the heap profiler and all added codepaths still require explicit
opt in via the extra define.

This series also includes some optimizations that especially improve
small allocations and hence we end up faster than the current state even with
the profiler on, under the assumption that small allocations (which here means
<= 16 KiB) are considerably more frequent than larger allocations, which is 
overwhelmingly the case for most applications.

For example, a new/delete pair for a small allocation prior to this series take on
average ~104 instructions, versus ~73 after _with heap profiling on_ or ~67 with
it compiled out. Performance measured in time rather than instructions generally
follows the same improvement factors as time, though with more variance.

To aid with analysis of the above we also added a set of microbenchmarks. Below
you will find the selected key numbers from this benchmark comparing current
master and this patch set with the profiler compiled in and
out. 

Note profiler-on numbers are taken from a run with very high sample period
to avoid actually taking any samples. Sample creation is fully dominated by
backtrace creation so this avoids noise.

With the profiler enabled at compile time the runtime performance is identical 
between "compiler turned off at runtime" and "very large sample period such
that zero samples are actually taken". That is, there is no additional check for
sampler enablement on the fast path, only the single check to see if the sampling
period has been reached.

There are a lot of profiling results, so I'll share first this summary of a single 
test, which uses `new` to allocate many small pointers and then deletes them.

The configurations show are:

*Baseline:* Current `master` without any of the changes in this series.
*New-OFF:* This series but with heap profiling disabled at compile time.
*New-ON:* This series and with heap profiling enabled at compile time.

First the instruction counts:

| Config | Compiler | test | instructions |
| - | - | - | -: | 
| Baseline | gcc-11   | alloc_bench.op_new_delete                | 97.6 |
| New-OFF  | gcc-11   | alloc_bench.op_new_delete                | 74.6 |
| New-ON   | gcc-11   | alloc_bench.op_new_delete                | 82.6 |
| Baseline | clang-16 | alloc_bench.op_new_delete                | 104.6 |
| New-OFF  | clang-16 | alloc_bench.op_new_delete                | 67.5|
| New-ON   | clang-16 |alloc_bench.op_new_delete                |  73.5 |

Next, runtimes on Skylake and Zen 4, these are from separate runs than the instruction counts above as I found that enabling instruction count measurement disturbed (i.e., increased) the timings by a significant amount, probably because some part of the perf syscall cost ends up timed.

| Config | Arch | Compiler | test                                     |  iterations |      median |         mad |         min |         max |
|-|-|-| -                                        |           - |           -: |           -: |           -: |           -: |
| Baseline | Skylake | clang-16  | alloc_bench.op_new_delete                |   134081000 |     7.376ns |     0.005ns |     7.370ns |     7.534ns |
| New-OFF  | Skylake | clang-16  | alloc_bench.op_new_delete                |   146836000 |     6.586ns |     0.008ns |     6.577ns |     6.765ns |
| New-ON   | Skylake | clang-16  | alloc_bench.op_new_delete                |   193241000 |     5.113ns |     0.004ns |     5.110ns |     5.246ns |
||
| Baseline | Zen4 | gcc-11   | alloc_bench.op_new_delete                |   207025000 |     4.796ns |     0.000ns |     4.795ns |     4.797ns |
| New-OFF  | Zen4 | gcc-11   | alloc_bench.op_new_delete                |   305103000 |     3.241ns |     0.000ns |     3.240ns |     3.241ns |
| New-ON   | Zen4 | gcc-11   | alloc_bench.op_new_delete                |   302127000 |     3.271ns |     0.000ns |     3.271ns |     3.273ns |
||
| Baseline | Zen4 | clang-16 | alloc_bench.op_new_delete                |   212207000 |     4.663ns |     0.000ns |     4.661ns |     4.663ns |
| New-OFF  | Zen4 | clang-16 | alloc_bench.op_new_delete                |   358628000 |     2.742ns |     0.001ns |     2.740ns |     2.745ns |
| New-ON   | Zen4 | clang-16 | alloc_bench.op_new_delete                |   347325000 |     2.833ns |     0.003ns |     2.830ns |     2.836ns |

The primary anomalous result is New-OFF on clang-16 on Skylake (also shows up on Alder Lake, not shown). The runtime is considerably slower than the instruction count would suggest (i.e., it is lower IPC than other configurations) and I don't see any problem in a manual inspection of the assembly. The same function is fast on Zen 2 and Zen 4. I believe it is probably an Intel-specific alignment effect. I did move the bottom of the test loop around to avoid the jcc erratum (results not shown) which improved the results but the result was still slower than expected. In any case, even this configuration is faster than the baseline and I expect the issue to go away at some point.

Large allocations are a bit slower, very slightly when heap profiling is off at compile time and more when they are enabled, e.g., here with 32,000 byte allocations:

*Baseline*
```
test                                      iterations      median         mad         min         max      allocs       tasks        inst
alloc_bench.alloc_free_large                72444000    13.125ns     0.010ns    13.073ns    13.140ns       1.000       0.000       336.1
```

*New-OFF*
```
test                                      iterations      median         mad         min         max      allocs       tasks        inst
alloc_bench.alloc_free_large                71368000    13.334ns     0.004ns    13.287ns    13.338ns       1.000       0.000       343.0
```

*New-ON*
```
test                                      iterations      median         mad         min         max      allocs       tasks        inst
alloc_bench.alloc_free_large                65224000    14.623ns     0.002ns    14.604ns    14.628ns       1.000       0.000       399.5
```

Such allocations are rare compared to small allocations, however, and in any case the relative cost of allocation goes down compared to the work done on the allocated memory itself (e.g., writing into it) and that ratio becomes ~negligible except for an application which allocates large regions but then does not use them.

We have many micro-benchmarks but this description is already quite long: let me know what you think would be useful.

### CODE REVIEW

**Comment by avikivity** [comment]
Looks very promising from reading the cover letter, but as this is delicate material I'd like multiple reviewers. Enlisting @michoecho

File: tests/perf/allocator_perf.cc
Comment (avikivity) [inline tests/perf/allocator_perf.cc:54]: please use 4-space indents throughout

Comment (avikivity) [reply tests/perf/allocator_perf.cc:54]: also 1 << 0, 2 << 0? Did you mean 1<< 0, 1 << 1?

File: tests/perf/allocator_perf.cc
Comment (avikivity) [inline tests/perf/allocator_perf.cc:60]: bool? what's M?

File: include/seastar/core/memory.hh
Comment (avikivity) [inline include/seastar/core/memory.hh:428]: It's not good for operator== to only compare part of the struct, but I see it's a preexisting problem so no need to fix it in the series.

File: src/core/memory.cc
Comment (avikivity) [inline src/core/memory.cc]: Please keep to the coding style which puts the brackets on the same line as the if.

File: src/core/memory.cc
Comment (avikivity) [inline src/core/memory.cc:444]: #ifdef HEAPPROF

File: include/seastar/util/sampler.hh
Comment (avikivity) [inline include/seastar/util/sampler.hh]: This won't work

File: src/core/memory.cc
Comment (avikivity) [inline src/core/memory.cc:1132]: Please fold into the patch that introduces the problem.

**Comment by avikivity** [comment]
Looks good. Please make newly introduced public APIs doxygen documented and verify the rendering.

File: include/seastar/util/sampler.hh
Comment (StephanDollberg) [reply include/seastar/util/sampler.hh]: You have probably realized at this point but I just want to clarify the approach we took here. 

We got advised that when importing only a part of a library it's best to make it clear what the original imported work is and what changes were made on top. 

Hence, we split the import & change into two. The first commit (what you commented on and what will definitely not compile) just imports the original file as is. The second commit then does the modifications.

If you prefer any other way of doing this we are of course very happy to change.

File: tests/perf/allocator_perf.cc
Comment (travisdowns) [reply tests/perf/allocator_perf.cc:54]: Let's say this was an easter egg to ensure reviewers were awake which will be removed on a subsequent push :).

File: src/core/memory.cc
Comment (travisdowns) [reply src/core/memory.cc:1132]: This is addressed in https://github.com/scylladb/seastar/commit/9d5d2445547b0eb47feef73a708dc9aeb1fbe005.

File: tests/perf/allocator_perf.cc
Comment (travisdowns) [reply tests/perf/allocator_perf.cc:54]: This is addressed (fix and offending commit folded) in https://github.com/scylladb/seastar/commit/9d5d2445547b0eb47feef73a708dc9aeb1fbe005.

File: include/seastar/util/sampler.hh
Comment (avikivity) [reply include/seastar/util/sampler.hh]: No, it's fine and a good approach.

File: tests/perf/allocator_perf.cc
Comment (avikivity) [reply tests/perf/allocator_perf.cc:54]: oh come on, the only thing that triggers me more is trailing whitespace

File: tests/perf/allocator_perf.cc
Comment (avikivity) [reply tests/perf/allocator_perf.cc:64]: No please change the name M to something meaningful. I know templates are supposed to be difficult but they do allow multi-character names.

And you could narrow the int to bool by writing something != 0.

File: include/seastar/util/sampler.hh
Comment (avikivity) [reply include/seastar/util/sampler.hh]: A line in the patch changelog would help (for reviewers that read the changelog).

Comment (StephanDollberg) [reply include/seastar/util/sampler.hh]: Have amended the commit text

File: src/core/memory.cc
Comment (StephanDollberg) [reply src/core/memory.cc:445]: Added

File: src/core/memory.cc
Comment (StephanDollberg) [reply src/core/memory.cc]: Fixed a bunch of those.

**Comment by StephanDollberg** [comment]
> Looks good. Please make newly introduced public APIs doxygen documented and verify the rendering.

I have added more of the doxygen parameters and checked the generated html.

**Comment by travisdowns** [comment]
@michoecho have you had a chance to take a peek at this one? All of @avikivity's feedback should be addressed.

File: tests/perf/allocator_perf.cc
Comment (travisdowns) [reply tests/perf/allocator_perf.cc]: After another look at this I was writing 20 lines to save 5 (numbers approximate), so I've refactored this a bit and the NTTP has disappeared and I think it's cleaner overall. Thanks for the eyes on this.

File: tests/perf/allocator_perf.cc
Comment (travisdowns) [reply tests/perf/allocator_perf.cc:56]: This reminded me to dust off my .clang-format patch for seastar. Get ready for some vigorous debate!

**Review by michoecho — APPROVED** [review]
Sorry for the delay, I forgot about this.

Looks good. I left a few comments, but feel free to ignore the nitpicks.

File: src/core/memory.cc
Comment (travisdowns) [reply src/core/memory.cc:452]: Yes, it declares `small_pool::allocate` to be inline, which increases the chance of it being inlined and also prevents an out-of-line definition from being included in the object file and binary.

File: include/seastar/util/sampler.hh
Comment (travisdowns) [reply include/seastar/util/sampler.hh:155]: The important part is that `maybe_sample` should use a `< 0` check instead of a `<= 0` check since this allows the whole decrement and compare to 0 operation which is on the hot path of every allocation to be serviced by a single RMW like `sub [interval_to_next_sample_], rax` followed by a comparison using the carry flag. IIRC that saved 3 instructions on the hot path though the net improvement is less than this would suggest as an RMW is more expensive than your average instruction (though less expensive than the replacement sequence).

To make that change correct, I had to decrement the initial value of `interval_to_next_sample_` by 1 which is why the  +1 disappeared.

Note that the change is from `next + 1` to `next` not the other way around (your comment suggests the other way around).

File: include/seastar/util/sampler.hh
Comment (travisdowns) [reply include/seastar/util/sampler.hh:154]: Interesting question, I'm not sure. This is probably from the original source the sampler was drawn from. @StephanDollberg do you know why this was used?

It wouldn't surprise me if exponential was faster as continuous distributions with closed formed floating point formulas often end up faster than integer formulas which need to calculate integer powers, etc.

File: include/seastar/core/memory.hh
Comment (travisdowns) [reply include/seastar/core/memory.hh:469]: Yeah possibly. What would you suggest if the check fails though? `abort()`?

File: tests/perf/allocator_perf.cc
Comment (travisdowns) [reply tests/perf/allocator_perf.cc:101]: "Single" in the sense that the tests above report the average cost of a single allocation (as they return the allocation count so the test can normalize by that value).

File: tests/unit/alloc_test.cc
Comment (travisdowns) [reply tests/unit/alloc_test.cc]: Good eye. I accept your offer to ignore this nitpick however as rewriting it to fix this will take a few minutes I could spend doing something like getting a the sesatar codebase auto-formatted. However if there is a concern about bisectability I can fix it.

File: src/core/memory.cc
Comment (travisdowns) [reply src/core/memory.cc]: Good eye. I accept your offer to ignore this nitpick however as rewriting it to fix this will take a few minutes I could spend doing something like improving sesatar compile times.

File: include/seastar/util/sampler.hh
Comment (StephanDollberg) [reply include/seastar/util/sampler.hh:154]: Yes exactly, the perfetto blog post has some numbers on that:

```
STL Distributions
Benchmarking of the STL distributions on a Pixel 4 reinforces our approach of estimating the geometric distribution using an exponential distribution, as its performance is ~8x better ([full results](https://gist.github.com/fmayer/3aafcaf58f8db09714ba09aa4ac2b1ac)).

Draw sample from exponential distribution with p = 1 / 32000:

ARM64: 21.3ns (sd 0.066ns, negligibly small, smaller than a CPU cycle)

ARM32: 33.2ns (sd 0.011ns, negligibly small, smaller than a CPU cycle)

Draw sample from geometric distribution with p = 1 / 32000:

ARM64: 169ns (sd 0.023ns, negligibly small, smaller than a CPU cycle) (7.93x)

ARM32: 222ns (sd 0.118ns, negligibly small, smaller than a CPU cycle) (6.69x)
```

source: https://perfetto.dev/docs/design-docs/heapprofd-sampling#performance-considerations

Not sure how this would look on beefier x86 processors.

File: tests/perf/allocator_perf.cc
Comment (michoecho) [reply tests/perf/allocator_perf.cc:101]: I see.

File: include/seastar/core/memory.hh
Comment (michoecho) [reply include/seastar/core/memory.hh:469]: Abort, yes.

If we think aborting would be too dangerous/unexpected to the poor user who attempts to use it in production without testing it, then we could put a warning in the one place a user is guaranteed to read — the name of the function. Like `assert_there_are_no_allocations_and_set_heap_profiling_sampling_rate.`

@avikivity What's your opinion on this?

File: tests/unit/alloc_test.cc
Comment (michoecho) [reply tests/unit/alloc_test.cc]: It probably is a bisectability concern, yes. But then again, if we really cared about bisectability, we would make the CI test all commits. Who knows how many bad commits we have out there.

@avikivity How much do we care about the bisectability here?

File: src/core/memory.cc
Comment (michoecho) [reply src/core/memory.cc]: > improving sesatar compile times

Modules will save us. R-right?

File: include/seastar/util/sampler.hh
Comment (michoecho) [reply include/seastar/util/sampler.hh:155]: If we go with this logic, then I think you should also change the inequality `while (interval_to_next_sample_ <= 0)` to `< 0`. (Otherwise, e.g. when you sample allocations of size 1 with sampling interval 1, a sampled allocation will always be followed by an unsampled one).

(Yes, of course this doesn't matter for any practical purpose).

File: include/seastar/core/memory.hh
Comment (travisdowns) [reply include/seastar/core/memory.hh:469]: Perhaps the comment also makes it sound scarier than it is. I should clarify that re-enabling sampling with a different rate doesn't cause any crash, corruption or undefined behavior: rather you are simply mixing samples with different frequencies and so your the aggregate samples will be biased, i.e., the samples which were taken when the sampling rate were lower should have a higher weight in principle but nothing like that is implement.

So the comment there from @StephanDollberg is indicating what you would need to do to avoid this bias (not have any allocations, which would also mean no samples), which of course is not really feasible in any case.

My recommendation would be that we soften the comment to detail the "problems" with mixing sampling of different rates but that we don't actually block it because it would still be useful in some situations. FWIW, the way we use the profiler currently in Redpanda is that is always on, and at a fixed rate.

WDYT?

File: src/core/memory.cc
Comment (travisdowns) [reply src/core/memory.cc]: I will play with modules from the retirement home I'll be in I guess.

File: include/seastar/util/sampler.hh
Comment (travisdowns) [reply include/seastar/util/sampler.hh:154]: I guess it depends heavily on the stdlib since with libstdc++ I measure them to have almost identical performance of ~66 ns for exponential and 70 ns for geometric. Looking and their implementations they are almost identical: both internally taking a floating point log of a random value.

In any case they should produce virtually identical results here?

File: tests/unit/alloc_test.cc
Comment (travisdowns) [reply tests/unit/alloc_test.cc]: @michoecho - in the force push https://github.com/scylladb/seastar/commit/a7d7f691117f8971d240937bb789cd3e9d8c87ed I have re-written this commit to move the test changes into a later commit and checked that the re-written commits all compile and pass the alloc test.

File: include/seastar/util/sampler.hh
Comment (travisdowns) [reply include/seastar/util/sampler.hh:155]: Good call! In force push https://github.com/scylladb/seastar/commit/a7d7f691117f8971d240937bb789cd3e9d8c87ed I have updated that loop to use `< 0`.

cc @StephanDollberg

File: src/core/memory.cc
Comment (travisdowns) [reply src/core/memory.cc:452]: > Does it? I was under the impression that `inline` is only for linkage nowadays.

It's the conventional wisdom in some sense that "it doesn't matter" but at least gcc definitely still uses it as input to its inlining heuristic: easy to show a case where it inlines only with inline.

Usually we get "implicit inline" by declaring the body inside the class but for whatever reason here the functions are not declared in the class body, but later on in the file so they lost the implicit inline.

File: include/seastar/core/memory.hh
Comment (travisdowns) [reply include/seastar/core/memory.hh]: @michoecho thanks for the feedback. I've adjusted this to use a raw pointer + length as you've suggested here: https://github.com/scylladb/seastar/pull/1904/commits/b0f4726151790d0d7adc9daaca80473413737ff4.

I've also added some testing of the raw interface.

**Comment by travisdowns** [comment]
@michoecho - I think all the feedback so far as been addressed, have a look when you have some time.

File: tests/unit/alloc_test.cc
Comment (tchaikov) [inline tests/unit/alloc_test.cc:522]:
```suggestion
    for (auto ptr : ptrs) {
```

**Review by tchaikov — COMMENTED** [review]
i didn't get a chance to review this changeset properly. just left some comments on formatting.

**Comment by travisdowns** [comment]
Force https://github.com/scylladb/seastar/commit/03e9966729c727d850795cc5facaa49f716cb1e3 was to update also the `sampled_memory_profile` stub method to the raw interface.

File: tests/unit/alloc_test.cc
Comment (StephanDollberg) [reply tests/unit/alloc_test.cc]: Fixed

File: tests/unit/alloc_test.cc
Comment (StephanDollberg) [reply tests/unit/alloc_test.cc]: Fixed, this particular one was actually unneeded.

File: include/seastar/core/memory.hh
Comment (StephanDollberg) [reply include/seastar/core/memory.hh]: I have added said check and updated the comment to be more clear. I have also added a test covering this scenario.

**Comment by travisdowns** [comment]
@tchaikov and @michoecho I think the most recent round of feedback has been addressed. If you can let us know if there is anything outstanding that we haven't addressed it would be much appreciated!

File: src/core/memory.cc
Comment (tchaikov) [inline src/core/memory.cc:1626]: nit, wrong indent.

**Comment by tchaikov** [comment]
> @tchaikov and @michoecho I think the most recent round of feedback has been addressed. If you can let us know if there is anything outstanding that we haven't addressed it would be much appreciated!

@travisdowns hi Travis, i think we are pending on a maintainer (hey @avikivity !) to review and merge this change. since maintainers are consistently been preempted by other tasks, guess a visible latency is expected =(

**Comment by avikivity** [comment]
I'm waiting for @michoecho's blessing

**Comment by michoecho** [comment]
> I'm waiting for @michoecho's blessing

Ouch, I didn't think you would wait for me again. The first review was already an approval, after all. I only left some nitpicks, they didn't deserve to block anything in the first place.

I can't see the changes since the last review (since github already GCd the hash I reviewed...), but I see that my comments have been addressed, so — assuming that no regression has sneaked in in the meantime — I (still) approve.

Sorry for stalling this PR for so long. I have no excuse — I should keep better track of review requests.

**Comment by tchaikov** [comment]
@avikivity hi Avi, ping?

**Comment by avikivity** [comment]
> > I'm waiting for @michoecho's blessing
> 
> Ouch, I didn't think you would wait for me again. The first review was already an approval, after all. I only left some nitpicks, they didn't deserve to block anything in the first place.
> 
> I can't see the changes since the last review (since github already GCd the hash I reviewed...), but I see that my comments have been addressed, so — assuming that no regression has sneaked in in the meantime — I (still) approve.
> 
> Sorry for stalling this PR for so long. I have no excuse — I should keep better track of review requests.

The trick is to push the blame on the contributor. @travisdowns please ping vigorously if your patches don't make progress.

File: include/seastar/core/memory.hh
Comment (avikivity) [reply include/seastar/core/memory.hh]: Looks like it's still using vector&.

We don't generally include a bug/deficiency and fix it in a later patch in the same set, but fold them together.

File: src/core/memory.cc
Comment (avikivity) [reply src/core/memory.cc:1626]: preexisting

File: include/seastar/core/memory.hh
Comment (avikivity) [inline include/seastar/core/memory.hh:459]: Please fold this into the original patch.

**Comment by travisdowns** [comment]
> The trick is to push the blame on the contributor.

:rofl: 

> @travisdowns please ping vigorously if your patches don't make progress.

Ack.

File: include/seastar/core/memory.hh
Comment (travisdowns) [reply include/seastar/core/memory.hh:459]: This is now folded in force push: https://github.com/scylladb/seastar/compare/58e4eb2dd3d67da55c0b76a186f045cd47ac3996..7a97ddfda2fec056975f00f20a8859c25bac41d9

**Comment by travisdowns** [comment]
@avikivity your feedback is addressed.

I've also pulled into one very small edge-case fix we've applied downstream as: https://github.com/scylladb/seastar/pull/1904/commits/a3548bcbbfa3ec93c04e7d8a65646072c8a35173

**Comment by travisdowns** [comment]
A big thank you for all the reviews and collaboration here. Hope this feature proves useful to you: it has already paid off nicely in Redpanda (the key is to emit some stacks in the OOM error message, which is up to the application currently).

File: src/core/memory.cc
Comment (tchaikov) [reply src/core/memory.cc:1629]: fixed at #2025

**Comment by avikivity** [comment]
This series was, in fact, mentioned a few ours ago on one of our OOM issues.

**Pull request merged**
